### PR TITLE
Small optimizations for analysis and planning

### LINKED ISSUE #5188: Planning time has increased sharply since 0.144
See Presto's planning time increased a lot since 0.144, especially for tables with large number of partitions, used to have only a few seconds planning time, now comes to minutes.

### PR DESCRIPTION
**Pull request by geraint0923**
Try to fix #5188

### CODE REVIEW

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java [hunk 1]
```diff
@@ -32,6 +32,9 @@
     private final List<String> parts;
     private final List<String> originalParts;
 
+    private Optional<QualifiedName> cachedPrefix;
+    private String cachedSuffix;
```
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java]: This class is meant to be immutable and thread-safe. This change breaks that assumption.

Comment (geraint0923) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java]: @martint  I could set these two variables as `final` and initialize them in its constructor.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java [hunk 2]
```diff
@@ -64,7 +68,12 @@ public QualifiedName(Iterable<String> parts)
     @Override
     public String toString()
     {
-        return Joiner.on('.').join(parts);
+        if (prefix.isPresent()) {
+            return prefix + suffix;
```
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java]: This is missing a `+ "." +`

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java [hunk 3]
```diff
@@ -73,27 +82,28 @@ public String toString()
...
-        if (parts.size() == 1) {
-            return Optional.empty();
-        }
-
-        return Optional.of(new QualifiedName(parts.subList(0, parts.size() - 1)));
+        return prefix;
     }
 
-    public boolean hasSuffix(QualifiedName suffix)
+    public boolean hasSuffix(QualifiedName suffixQualifiedName)
```
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java]: Why did you rename this variable? `suffix` was good

Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java]: Oh, I see... it makes the usages below look weird (e.g., `suffix.suffix`). I'd just call it `name`, then.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java [hunk 4]
```diff
@@ -47,13 +48,16 @@ public QualifiedName(Iterable<String> parts)
     {
         requireNonNull(parts, "parts is null");
         checkArgument(!isEmpty(parts), "parts is empty");
-        this.parts = ImmutableList.copyOf(transform(parts, part -> part.toLowerCase(ENGLISH)));
+        List<String> lowerParts = ImmutableList.copyOf(transform(parts, part -> part.toLowerCase(ENGLISH)));
+
         this.originalParts = ImmutableList.copyOf(parts);
+        this.prefix = lowerParts.size() == 1 ? Optional.empty() : Optional.of(new QualifiedName(lowerParts.subList(0, lowerParts.size() - 1)));
```
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java]: This expression is a bit hard to read. Use a normal `if`

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java (hunk 3 above)
Comment (geraint0923) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java]: Nice, I just don't want to use a duplicate variable name like `suffix`, `name` is pretty good.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/TypeSignature.java [hunk 5]
```diff
@@ -281,7 +281,7 @@ else if (c == ')') {
         verify(parameters.size() == fieldNames.size() || fieldNames.isEmpty(), "Number of parameters and fieldNames for ROW type doesn't match");
         List<TypeSignatureParameter> result = new ArrayList<>();
         for (int i = 0; i < parameters.size(); i++) {
-            String fieldName = fieldNames.isEmpty() ? format("field%d", i) : fieldNames.get(i);
+            String fieldName = fieldNames.isEmpty() ? "field" + i : fieldNames.get(i);
```
Comment (haozhun) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/TypeSignature.java]: For things like this, you need to add comments so that people don't turn this back into `format`.

**Comment by geraint0923** [comment]
Address the comments.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/TypeSignature.java [hunk 6]
```diff
@@ -281,7 +281,8 @@ else if (c == ')') {
         verify(parameters.size() == fieldNames.size() || fieldNames.isEmpty(), "Number of parameters and fieldNames for ROW type doesn't match");
         List<TypeSignatureParameter> result = new ArrayList<>();
         for (int i = 0; i < parameters.size(); i++) {
-            String fieldName = fieldNames.isEmpty() ? format("field%d", i) : fieldNames.get(i);
+            // Use `"field" + i` instead of `String.format` to avoid expensive string formatting
```
Comment (martint) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/TypeSignature.java]: Remove the backticks from the comment

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java [hunk 7]
```diff
@@ -64,7 +73,12 @@ public QualifiedName(Iterable<String> parts)
     @Override
     public String toString()
     {
-        return Joiner.on('.').join(parts);
+        if (prefix.isPresent()) {
+            return prefix.get() + "." + suffix;
+        }
+        else {
```
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java]: else is redundant

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java [hunk 8]
```diff
@@ -73,27 +87,28 @@ public String toString()
...
-        if (parts.size() == 1) {
-            return Optional.empty();
-        }
-
-        return Optional.of(new QualifiedName(parts.subList(0, parts.size() - 1)));
+        return prefix;
     }
 
-    public boolean hasSuffix(QualifiedName suffix)
+    public boolean hasSuffix(QualifiedName name)
```
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java]: While you're at it, can you add some tests to make sure the current behavior didn't break?

**Comment by martint** [comment]
Remove the backticks from the commit messages

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 9]
```diff
@@ -150,6 +150,23 @@ public void testPossibleExponentialBacktracking()
     }
 
     @Test
+    public void testQualifiedName()
+    {
+        QualifiedName name = QualifiedName.of("a", "b", "c", "d");
```
Comment (martint) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: I'd just inline all these variables. They are simple assertions, and it would be easier to read if they read like:

```java
assertTrue(QualifiedName.of("a", "b", "c", "d").hasSuffix(QualifiedName.of("c", "d"));
assertFalse(QualifiedName.of("a", "b", "c", "d").hasSuffix(QualifiedName.of("a", "b"));
...
```

**Comment by martint** [comment]
Other than minor comments, it looks good.

**Comment by geraint0923** [comment]
Addressed the comment.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/SimplePlanRewriter.java [hunk 10]
```diff
@@ -84,7 +83,9 @@ public PlanNode defaultRewrite(PlanNode node, C context)
         public PlanNode rewrite(PlanNode node, C userContext)
         {
             PlanNode result = node.accept(nodeRewriter, new RewriteContext<>(nodeRewriter, userContext));
-            requireNonNull(result, format("nodeRewriter returned null for %s", node.getClass().getName()));
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/SimplePlanRewriter.java]:
```java
verify(result != null, "nodeRewriter returned null for %s", node.getClass().getName());

File: presto-main/src/main/java/com/facebook/presto/type/TypeUtils.java [hunk 11]
```diff
@@ -105,7 +104,11 @@ public static boolean positionEqualsPosition(Type type, Block leftBlock, int lef
 
     public static Type resolveType(TypeSignature typeName, TypeManager typeManager)
     {
-        return requireNonNull(typeManager.getType(typeName), format("Type '%s' not found", typeName));
+        Type type = typeManager.getType(typeName);
+        if (type == null) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/type/TypeUtils.java]:
```java
checkArgument(type != null, "Type '%s' not found", typeName);

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java [hunk 12]
```diff
@@ -45,10 +45,21 @@ public QualifiedName(String name)
 
     public QualifiedName(Iterable<String> parts)
     {
+        this(parts, true);
+    }
+
+    private QualifiedName(Iterable<String> parts, boolean toLower)
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java]: I don't like this, but I'm having trouble coming up with a better way to write this. @martint @dain any ideas?

File: presto-spi/src/main/java/com/facebook/presto/spi/type/TypeSignature.java [hunk 13]
```diff
@@ -282,7 +282,8 @@ else if (c == ')') {
         verify(parameters.size() == fieldNames.size() || fieldNames.isEmpty(), "Number of parameters and fieldNames for ROW type doesn't match");
         List<TypeSignatureParameter> result = new ArrayList<>();
         for (int i = 0; i < parameters.size(); i++) {
-            String fieldName = fieldNames.isEmpty() ? format("field%d", i) : fieldNames.get(i);
+            // Use "field" + i instead of String.format to avoid expensive string formatting
```
Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/TypeSignature.java]: No need for this comment

File: presto-spi/src/main/java/com/facebook/presto/spi/type/TypeSignature.java [hunk 14]
```diff
@@ -282,7 +282,8 @@ else if (c == ')') {
...
+            String fieldName = fieldNames.isEmpty() ? "field" + i : fieldNames.get(i);
```
Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/TypeSignature.java]: Add parens around `"field" + i`

File: presto-spi/src/main/java/com/facebook/presto/spi/type/TypeSignature.java (hunk 13 above)
Comment (geraint0923) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/TypeSignature.java]: Add this comment just in case someone change this back to `format(..)` in the future.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java (hunk 12 above)
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/QualifiedName.java]: Maybe a better way would be to have a private constructor that just takes the parts as they come (so it can be used in getPrefix), and a public static factory method that first downcases them before creating the object. It would require changing all call sites of `new QualifiedName(...)`, though.

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 15]
```diff
@@ -150,6 +150,18 @@ public void testPossibleExponentialBacktracking()
     }
 
     @Test
+    public void testQualifiedName()
+    {
+        assertEquals(QualifiedName.of("a", "b", "c", "d").toString(), "a.b.c.d");
+        assertEquals(QualifiedName.of("A", "b", "C", "d").toString(), "a.b.c.d");
+        assertEquals(QualifiedName.of("a", "b", "c", "d").hasSuffix(QualifiedName.of("b", "c", "d")), true);
```
Comment (martint) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: Use assertTrue/assertFalse for boolean tests.

**Comment by geraint0923** [comment]
@martint I just updated the PR according to what we discussed. Could you please take a look?

**Comment by martint** [comment]
I fixed a couple logical conflicts due to recent changes and merged it
